src/prss.rs

```diff
@@ -0,0 +1,493 @@
+
+impl ParticipantSetup {
+    const CONTEXT_VALUES: &'static [u8] = b"IPA PRSS values";
+    const CONTEXT_BITS: &'static [u8] = b"IPA PRSS bits";
+
+    /// Construct a new, unconfigured participant.  This can be configured by
+    /// providing public keys for the left and right participants to `setup()`.
+    pub fn new<R: RngCore + CryptoRng>(r: &mut R) -> Self {
+        Self {
+            left: KeyExchange::new(r),
+            right: KeyExchange::new(r),
+        }
+    }
+
+    /// Get the public keys that this setup instance intends to use.
+    /// The public key that applies to the left participant is at `.0`,
+    /// with the key for the right participant in `.1`.
+    #[must_use]
+    pub fn public_keys(&self) -> (PublicKey, PublicKey) {
+        (self.left.public_key(), self.right.public_key())
+    }
+
+    /// Provide the left and right public keys to construct a functioning
+    /// participant instance.
+    #[must_use]
+    pub fn setup(self, left_pk: &PublicKey, right_pk: &PublicKey) -> Participant {
+        let fl = self.left.key_exchange(left_pk);
+        let fr = self.right.key_exchange(right_pk);
+        Participant {
+            left: fl.generator(Self::CONTEXT_VALUES),
+            left_bits: BitGenerator::from(fl.generator(Self::CONTEXT_BITS)),
+            right: fr.generator(Self::CONTEXT_VALUES),
+            right_bits: BitGenerator::from(fr.generator(Self::CONTEXT_BITS)),
+        }
+    }
+}
+
+/// The key exchange component of a participant.
+pub struct KeyExchange {
+    sk: EphemeralSecret,
+}
+
+impl KeyExchange {
+    pub fn new<R: RngCore + CryptoRng>(r: &mut R) -> Self {
+        let mut r = rng::Adapter::from(r);
+        Self {
+            sk: EphemeralSecret::new(&mut r),
+        }
+    }
+
+    #[must_use]
+    pub fn public_key(&self) -> PublicKey {
+        PublicKey::from(&self.sk)
+    }
+
+    #[must_use]
+    pub fn key_exchange(self, pk: &PublicKey) -> GeneratorFactory {
+        debug_assert_ne!(pk, &self.public_key(), "self key exchange detected");
+        let secret = self.sk.diffie_hellman(pk);
+        let kdf = Hkdf::<Sha256>::new(None, secret.as_bytes());
+        GeneratorFactory { kdf }
+    }
+}
+
+/// This intermediate object exists so that multiple generators can be constructed,
+/// with each one dedicated to one purpose.
+pub struct GeneratorFactory {
+    kdf: Hkdf<Sha256>,
+}
+
+impl GeneratorFactory {
+    /// Create a new generator using the provided context string.
+    /// # Panics
+    /// Never: we don't let that happen.
+    #[must_use]
+    pub fn generator(&self, context: &[u8]) -> Generator {
+        let mut k = GenericArray::default();
+        self.kdf.expand(context, &mut k).unwrap();
+        Generator {
+            cipher: Aes256::new(&k),
+        }
+    }
+}
+
+/// The basic generator.  This generates values based on an arbitrary index.
+pub struct Generator {
+    cipher: Aes256,
+}
+
+impl Generator {
+    /// Generate the value at the given index.
```

Comment:
I don't have the full context here, how the values of `index` are picked by the clients of this API? do they set an initial value and then just monotonically increase it or pick at random using a PRNG?

Reply:
Drawing at random is generally a poor choice. I was thinking that the number would be structured.  You have some number of bits dedicated to a record counter, some number of bits dedicated to a step number, or something like that.

There might need to be a layer on top that makes this part a bit easier to manage, I don't know.